### PR TITLE
[Enhancement] only collect io of selected subfield of complex type

### PR DESCRIPTION
**Pull request by zombee0**
Fixes #issue
for complex type, we have pruned subfield, but collect_io_range collect all subfield,
this pr will enhance it by only collect io of selected subfield. 

## What type of PR is this:
- [ ] BugFix
- [ ] Feature
- [x] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

## Checklist:
- [x] I have added test cases for my bug fix or my new feature
- [ ] This pr will affect users' behaviors
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [x] 3.0
  - [x] 2.5
  - [ ] 2.4
  - [ ] 2.3

### CODE REVIEW

File: be/src/formats/parquet/column_reader.cpp [hunk 1]
```diff
@@ -483,6 +483,80 @@ class StructColumnReader : public ColumnReader {
...
+            continue;
+        }
+
+        int32_t field_id = iceberg_it->second->field_id;
+
+        auto parquet_field_it = field_id_2_pos.find(field_id);
+        if (parquet_field_it == field_id_2_pos.end()) {
+            // Means newly added struct subfield not existed in original parquet file, we put nullptr
+            // column reader in children_reader, we will append default value for this subfield later.
+            LOG(WARNING) << "Struct subfield name: " + format_subfield_name + " not found in ParquetField.";
```
Comment (DorianZheng) [inline be/src/formats/parquet/column_reader.cpp:549]: This is expected case, don't print warning logs.

**Review by DorianZheng — APPROVED** [review]

File: be/src/formats/parquet/group_reader.h [hunk 2]
```diff
@@ -133,8 +133,12 @@ class GroupReader {
 
     Status _read(const std::vector<int>& read_columns, size_t* row_count, ChunkPtr* chunk);
     Status _lazy_skip_rows(const std::vector<int>& read_columns, const ChunkPtr& chunk, size_t chunk_size);
-    void _collect_field_io_range(const ParquetField& field, std::vector<io::SharedBufferedInputStream::IORange>* ranges,
-                                 int64_t* end_offset);
+    void _collect_field_io_range(const ParquetField& field, const TypeDescriptor& col_type,
+                                 std::vector<io::SharedBufferedInputStream::IORange>* ranges, int64_t* end_offset);
+    void _collect_field_io_range_schema_change(const ParquetField& field, const TypeDescriptor& col_type,
```
[the remark this answers is not in the document]
Comment (zombee0) [reply be/src/formats/parquet/group_reader.h:138]: ok

File: be/src/formats/parquet/column_reader.cpp (hunk 1 above)
Comment (zombee0) [reply be/src/formats/parquet/column_reader.cpp:549]: will do it

**Review by Youngwb — DISMISSED** [review]

File: be/src/formats/parquet/column_reader.cpp [hunk 3]
```diff
@@ -511,34 +585,18 @@ Status ColumnReader::create(const ColumnReaderOptions& opts, const ParquetField*
...
         for (size_t i = 0; i < col_type.children.size(); i++) {
-            const std::string& formatted_subfield_name =
-                    opts.case_sensitive ? col_type.field_names[i]
-                                        : boost::algorithm::to_lower_copy(col_type.field_names[i]);
-
-            auto it = field_name_2_pos.find(formatted_subfield_name);
-            if (it == field_name_2_pos.end()) {
-                LOG(WARNING) << "Struct subfield name: " + formatted_subfield_name + " not found.";
-                // Put nullptr ColumnReader for non existed subfield, it will append_default when fill data from parquet.
+            if (subfield_pos[i] == -1) {
```
Comment (Smith-Cruise) [inline be/src/formats/parquet/column_reader.cpp:593]: add comment, like : `-1 means subfield not existed, we still need to emplace_back nullptr`

File: be/src/formats/parquet/column_reader.cpp [hunk 4]
```diff
@@ -588,52 +646,21 @@ Status ColumnReader::create(const ColumnReaderOptions& opts, const ParquetField*
...
-                    opts.case_sensitive ? each.name : boost::algorithm::to_lower_copy(each.name);
-            subfield_name_2_field_schema.emplace(format_subfield_name, &each);
-        }
-
-        std::unordered_map<int32_t, size_t> field_id_2_pos{};
-        for (size_t i = 0; i < field->children.size(); i++) {
-            field_id_2_pos.emplace(field->children[i].field_id, i);
-        }
+        std::vector<int32> subfield_pos(col_type.children.size());
+        std::vector<const TIcebergSchemaField*> iceberg_schema_subfield(col_type.children.size());
```
Comment (Smith-Cruise) [inline be/src/formats/parquet/column_reader.cpp:650]: `new_iceberg_schema_field` instead of `iceberg_schema_subfield`?

Comment (Smith-Cruise) [reply be/src/formats/parquet/column_reader.cpp:650]: And I think maybe we don't need to copy a `std::vector<const TIcebergSchemaField*>`, just use the original one.

File: be/src/formats/parquet/column_reader.h [hunk 5]
```diff
@@ -52,6 +52,16 @@ class ColumnReader {
     static Status create(const ColumnReaderOptions& opts, const ParquetField* field, const TypeDescriptor& col_type,
                          const TIcebergSchemaField* iceberg_schema_field, std::unique_ptr<ColumnReader>* output);
 
+    // for struct type without schema change
+    static void get_subfield_pos_with_pruned_type(const ParquetField& field, const TypeDescriptor& col_type,
+                                                  bool case_sensitive, std::vector<int32>& pos);
```
Comment (Smith-Cruise) [inline be/src/formats/parquet/column_reader.h:57]: `int32_t` instead `int32`?

**Review by dirtysalt — DISMISSED** [review]

**Review by dirtysalt — APPROVED** [review]

**Review by Youngwb — APPROVED** [review]

**Comment by wanpengfei-git** [comment]
@Mergifyio backport branch-3.0

**Comment by wanpengfei-git** [comment]
@Mergifyio backport branch-2.5

**Comment by zombee0** [comment]
https://github.com/Mergifyio backport branch-3.1

**Pull request merged**